### PR TITLE
Add a GetStoreStats() method to EtcdClient.

### PR DESCRIPTION
**Pull request by AlCutter**
Exposes the json formatted stats under etcd's "/v2/stats/store" key.

### CODE REVIEW

File: cpp/util/etcd.cc [hunk 1]
```diff
@@ -756,7 +804,18 @@ void EtcdClient::Delete(const string& key, const int64_t current_index,
...
+  Generic(key, kKeysSpace, params, UrlFetcher::Verb::DELETE, gen_resp, task);
+}
+
+
+void EtcdClient::GetStoreStats(StatsResponse* resp, Task* task) {
+  map<string, string> params;
+  GenericResponse* const gen_resp(new GenericResponse);
+  task->DeleteWhenDone(gen_resp);
+
+  Generic("/store", kStatsSpace, params, UrlFetcher::Verb::GET, gen_resp,
```
Comment (pphaneuf) [inline cpp/util/etcd.cc]: Make a constant for that "key"?

File: cpp/util/etcd.cc [hunk 2]
```diff
@@ -180,6 +191,42 @@ void GetRequestDone(const string& keyname, EtcdClient::GetResponse* resp,
...
 
 
+void CopyStat(const string& key, const JsonObject& from,
+              map<string, int64_t>* to) {
+  CHECK_NOTNULL(to);
+  const JsonInt stat(from, key.c_str());
+  if (!stat.Ok()) {
+    LOG(WARNING) << "Failed to find stat " << key;
+  }
+  (*to)[key] = stat.Value();
```
Comment (pphaneuf) [inline cpp/util/etcd.cc]: Should you set it if it doesn't exist? I'm not sure what this does with a null...

File: cpp/util/etcd.cc [hunk 3]
```diff
@@ -180,6 +191,42 @@ void GetRequestDone(const string& keyname, EtcdClient::GetResponse* resp,
...
+
+
+void GetStoreStatsRequestDone(EtcdClient::StatsResponse* resp,
+                              Task* parent_task,
+                              EtcdClient::GenericResponse* gen_resp,
+                              Task* task) {
+  *resp = EtcdClient::StatsResponse();
+  resp->etcd_index = gen_resp->etcd_index;
+  if (!task->status().ok()) {
+    parent_task->Return(Status(task->status().CanonicalCode(),
```
Comment (pphaneuf) [inline cpp/util/etcd.cc]: If you're not modifying the status, you should just return it directly:

    parent_task->Return(task->status());

File: cpp/util/fake_etcd.cc [hunk 4]
```diff
@@ -178,26 +181,37 @@ void FakeEtcdClient::InternalPut(const string& key, const string& value,
...
+}
+
+
 void FakeEtcdClient::Create(const string& key, const string& value,
                             Response* resp, Task* task) {
-  return InternalPut(key, value, system_clock::time_point::max(), true, -1,
-                     resp, task);
+  InternalPut(key, value, system_clock::time_point::max(), true, -1, resp,
+              task);
+  UpdateOperationStats("create", task);
```
Comment (pphaneuf) [inline cpp/util/fake_etcd.cc]: This isn't safe, the ```task``` could have already been deleted... You could register it as a cleanup callback on the task, though, those get run after ```Return``` has been called, but before the task is gone, so you could check the status on it. Plus, you ensure it getting called, no matter what crazy code path is taken.

File: cpp/util/etcd.cc (hunk 1 above)
Comment (AlCutter) [inline cpp/util/etcd.cc]: done

File: cpp/util/fake_etcd.cc [hunk 5]
```diff
@@ -178,26 +181,37 @@ void FakeEtcdClient::InternalPut(const string& key, const string& value,
 }
 
 
+void FakeEtcdClient::UpdateOperationStats(const string& op, const Task* task) {
+  CHECK_NOTNULL(task);
+  if (!task->IsActive()) {
```
Comment (pphaneuf) [inline cpp/util/fake_etcd.cc]: I'd fold the ```CHECK_NOTNULL``` unto this line, personally. Your call...

File: cpp/util/etcd.cc (hunk 2 above)
Comment (AlCutter) [inline cpp/util/etcd.cc]: whoops, missing `return` there, thanks.

File: cpp/util/etcd.cc (hunk 3 above)
Comment (AlCutter) [inline cpp/util/etcd.cc]: Heh, this was copied from `GetRequestDone`, and then later modified.
Fixed.

**Comment by pphaneuf** [comment]
Done for now, and your first update looks good, just the fake to fix up...

File: cpp/util/fake_etcd.cc (hunk 4 above)
Comment (AlCutter) [inline cpp/util/fake_etcd.cc]: Nice, thanks, done.

File: cpp/util/fake_etcd.cc (hunk 5 above)
Comment (AlCutter) [inline cpp/util/fake_etcd.cc]: I kinda prefer them separated out; scans a bit nicer I think.

File: cpp/util/fake_etcd.cc [hunk 6]
```diff
@@ -178,26 +181,40 @@ void FakeEtcdClient::InternalPut(const string& key, const string& value,
...
+  }
+}
+
+
 void FakeEtcdClient::Create(const string& key, const string& value,
                             Response* resp, Task* task) {
-  return InternalPut(key, value, system_clock::time_point::max(), true, -1,
-                     resp, task);
+  task->CleanupWhenDone(
+      std::bind(&FakeEtcdClient::UpdateOperationStats, this, "create", task));
```
Comment (pphaneuf) [inline cpp/util/fake_etcd.cc]: Take off these ```std::```, this is not an -inl.h file, there are rules! :wink:

**Comment by pphaneuf** [comment]
Other than those spurious ```std::```, LGTM.

File: cpp/util/fake_etcd.cc (hunk 6 above)
Comment (AlCutter) [inline cpp/util/fake_etcd.cc]: Rules, I think you'll find, are there to be broken.

**Pull request merged**